src/Symfony/Component/Dsn/ConnectionFactory.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Dsn;
+
+use Symfony\Component\Dsn\Exception\InvalidArgumentException;
+use Symfony\Component\Dsn\Factory\MemcachedConnectionFactory;
+use Symfony\Component\Dsn\Factory\RedisConnectionFactory;
+
+/**
+ * Factory for undetermined Dsn.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class ConnectionFactory
+{
+    const TYPE_REDIS = 1;
+    const TYPE_MEMCACHED = 2;
+
+    /**
+     * @param string $dsn
+     *
+     * @return int
+     */
+    public static function getType($dsn)
+    {
+        if (!is_string($dsn)) {
+            throw new InvalidArgumentException(sprintf('The %s() method expect argument #1 to be string, %s given.', __METHOD__, gettype($dsn)));
+        }
+        if (0 === strpos($dsn, 'redis://')) {
+            return static::TYPE_REDIS;
+        }
+        if (0 === strpos($dsn, 'memcached://')) {
+            return static::TYPE_MEMCACHED;
+        }
```

Comment:
this implementation is hard to extend / maintainance
when we will want to have 3rd and 4th dsn type, you would just add more if's, but when one wants to use component from outside, he won't have nice way to add more guessers.

what about some indicator classes created and registered in factory class as a list, each of them specialized in recognizing only one type and answering "yes" or "no" ? then, here we would just iterate over the list

Reply:
Instead of hardcoding a list of supported Factories, an other solution would be ot use instances (instead of static calls), definig a common interface with 2 methods `supports` and `createConnection(mixed, array): mixed` and use a strategy pattern to iterrate over a list of injected factories.
- do we really need OOP in factories?
- interfaces with mixed/mixed seems useless

Reply:
then at least mark this class as `final`, maybe even `@internal`

Reply:
Adding guesser on user's side is not needed: these are factories, they don't adhere to any contract, exactly like constructors. Instead, they ship behaviors, that user would then wire in their DI layer.
If they have more DSN factories, they'll use these factories directly. No need to create any OOP design here, that'd we completely useless and wouldn't solve any problem, just create new ones.

Reply:
I red your comments from 15 minutes ago. and especially for that:
> then at least mark this class as `final`, maybe even `@internal`

Reply:
`@internal` would defeat the purpose of the component by stating "don't use it" to users.